File: LRUCache.js
class Node {
    constructor(key, val) {
        this.key = key;
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}


class DLL {
    constructor() {
        this.size = 0;
        this.head = null;
        this.tail = null;
    }

    //to the end
    add(node) {

        if(this.head == null) {
            this.head = node;
            this.tail = this.head;
            this.size++;
            return;
        }
        
        this.tail.next = node;
        node.prev = this.tail;
        this.tail = node;
        this.size++;
    }

    //from begenning
    remove() {
        let temp = this.head;
        this.head = this.head.next;
        temp.next = null;
        this.size--;
        return temp;
    }
}


class LRUCache {
    map = {};
    dll = new DLL();

    constructor(capacity) {
        this.capacity = capacity;
    }

    get(key) {
        
    }

    put(key, value) {
        let currentSize = Object.keys(this.map).length;
        let newNode = new Node(key, value);

        if(this.map[key]!=undefined) {
          //remove that node

            
        }

        else if(currentSize < this.capacity) {
            this.dll.add(newNode);
        } else {
            let leastRecentlyUsed = this.dll.remove();
            delete this.map[leastRecentlyUsed.key];
            this.dll.add(newNode);
        }
        this.map[key] = newNode;
        
    }
    
}

File: LRUCache.java
class Node {
    int key;
    int value;
    Node prev;
    Node next;

    Node(int key, int value) {
        this.key = key;
        this.value = value;
    }
}

class LRUCache {
    Map<Integer,Node> map = new HashMap<>(); 
    int capacity;
    int currentSize = 0;
    Node MRU;
    Node LRU;


    public LRUCache(int capacity) {
        this.capacity = capacity;
    }
    
    public int get(int key) {
        if(map.containsKey(key)) {
            Node node =  map.get(key);
            updateMRU(node);
            return node.value;
        }
        return -1;
    }
    
    public void put(int key, int value) {
        Node node;

        //key already present, then update
        if(map.containsKey(key)) {
            node = map.get(key);
            node.value = value;
            updateMRU(node);
        } else {
            node = new Node(key, value);
            if(currentSize == capacity) {
                //evict
                Node temp = LRU;
                LRU = LRU.next;
                temp.next = null;
                if(LRU != null)
                    LRU.prev = null;
                currentSize--;
                map.remove(temp.key);
            }

            map.put(key, node);
            currentSize++;
            if(currentSize ==1) {
                LRU = node;
                MRU = node;
                return;
            }
            
            MRU.next = node;
            node.prev = MRU;
            MRU = node;
        }
    }

    private void updateMRU(Node node) {
        if(MRU == node) return;
        if(LRU == node) LRU = LRU.next;

        if(node.prev != null) {
            node.prev.next = node.next;
        }
        if(node.next != null) {
            node.next.prev = node.prev;
        }
        
        MRU.next = node;
        node.prev = MRU;
        node.next = null;
        MRU = node;
        
    }
}
